src/util/recalls.ts

```diff
@@ -0,0 +1,96 @@
+import { BearerResponse, MotRecalls, MotSecret } from '../models/motRecalls';
+import logger from './logger';
+
+/**
+ * Search retrieved recall data for an active recall then construct return object.
+ * @param vehicleRecalls
+ * @returns
+ */
+export const filterMotRecalls = (vehicleRecalls: MotRecalls) => {
+  logger.debug('Filter Recall Response');
+  const foundRecall = vehicleRecalls.recalls.find((recall) => {
+    if (recall.repairStatus === 'NOT_FIXED' && Date.parse(recall.recallCampaignStartDate) < Date.now()) {
+      return recall;
+    }
+    return false;
+  });
+  return {
+    manufacturer: foundRecall ? vehicleRecalls.manufacturer : null,
+    hasRecall: !!foundRecall,
+  };
+};
+
+/**
+ * Retrieve vehicle recall data from MOT recall API
+ * @param vin - vin is query parameter
+ * @returns Promise<motRecalls> - vehicle recall information
+ */
+export const getMotRecallsByVin = async (vin: string, cache: Map<string, string>, motSecret: MotSecret): Promise<MotRecalls | undefined> => {
+  logger.debug('Calling MOT Recalls');
+  try {
+    const bearerToken = cache.get('bearerToken') as string;
+    const motApiUrl = `${motSecret.apiURL}recalls/${vin}`;
+
+    let recallResponse = await fetch(motApiUrl, {
+      headers: {
+        Authorization: `Bearer ${bearerToken}`,
+        'X-API-Key': motSecret.apiKey,
+      },
+    });
+
+    logger.debug(`first recall response: ${JSON.stringify(recallResponse)}`);
+
+    if (recallResponse.status === 403 || recallResponse.status === 401) {
+      const newBearerToken = await getBearerToken(motSecret);
+      if (!newBearerToken) {
+        return undefined;
+      }
+      logger.debug('got a new bearer token');
+
+      cache.set('bearerToken', newBearerToken);
+      recallResponse = await fetch(motApiUrl, {
+        headers: {
+          Authorization: `Bearer ${bearerToken}`,
+          'X-API-Key': motSecret.apiKey,
+        },
+      });
+      logger.debug(`second recall response if called: ${JSON.stringify(recallResponse)}`);
+    }
```

Comment:
I wonder if it makes sense to change what we're caching based on this retry/unsuccessful flow.

By caching the entire `BearerResponse` object, we could avoid a number of bad requests ever hitting the MOT service by using the expiry of the token after decoding?

This would then mean we could determine whether to get a new token, or whether to use the existing one. We could introduce a ~5s buffer tolerance for the refresh.

That means the cache would become something like `const cache: Map<string, BearerResponse> = new Map();`?

What are your thoughts?

Reply:
Yeah I thought about this, I decided against it basically as I thought it was overkill. Odds are with a 35 min expire time, that a new instance will probably be created within that time. Also means considering we only have stuff happening between 8-5 we could only ever expire a token 15 times a day anyway. This also covers that if you get a random failure for some reason you get another retry at least once. 

What do you reckon? I could change this over but do you think it's worth it? Happy to do it if you think so

Reply:
I've always been used to tokens lasting ~15 mins, so having 35 mins TTL for the token does change things slightly, and you make a good point as you would think another instance would have been spun up within that time. 

Would it be a lot more work to add it in? I think anything to reduce unnecessary outbound calls is a good thing?

Otherwise, we could always go with this for now, make sure we are logging the errors and can improve further if we do we find we are getting a lot of expired token errors being sent back.